adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,546 @@
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_035: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, TLSIO_APPLEIOS_STATE_OPENING, TLSIO_APPLEIOS_STATE_OPEN, or TLSIO_APPLEIOS_STATE_CLOSING, the tlsio_appleios_open shall set the tlsio state as TLSIO_APPLEIOS_STATE_ERROR, and return _LINE_. ]*/
+            LogError("Try to open a connection with an already opened TLS.");
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+            result = __FAILURE__;
+        }
+        else
+        {
+		    CFStreamStatus status;
+			
+			status = (tlsio_instance->sockRead == NULL)? kCFStreamStatusClosed : CFReadStreamGetStatus(tlsio_instance->sockRead);
+			
+			if (status != kCFStreamStatusClosed)
+            {
+                /* Codes_SRS_TLSIO_APPLEIOS_32_037: [ If the ssl client is connected, the tlsio_appleios_open shall change the state to TLSIO_APPLEIOS_STATE_ERROR, log the error, and return _LINE_. ]*/
+                LogError("SSL stream is not closed.");
+                tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+                result = __FAILURE__;
+            }
+			else
+			{
+	            CFStreamCreatePairWithSocketToHost(NULL, tlsio_instance->remote_host, tlsio_instance->port, &tlsio_instance->sockRead, &tlsio_instance->sockWrite);
+				
+				if (tlsio_instance->sockRead == NULL || tlsio_instance->sockWrite == NULL)
+				{
+					LogError("Unable to create socket streams.")
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+				else if (!CFReadStreamSetProperty(tlsio_instance->sockRead, kCFStreamPropertySSLSettings, kCFStreamSocketSecurityLevelTLSv1))
+				{
+					LogError("Unable to start SSL on socket.");
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+	            else if (!CFReadStreamOpen(tlsio_instance->sockRead) || !CFWriteStreamOpen(tlsio_instance->sockWrite))
+				{
+					LogError("Failure: connect failure");
+					CFReadStreamClose(tlsio_instance->sockRead);
+					CFWriteStreamClose(tlsio_instance->sockWrite);
+					tlsio_instance->sockRead = NULL;
+					tlsio_instance->sockWrite = NULL;
+					result = __FAILURE__;
+				}
+				else
+				{
+					tlsio_instance->countTry = MAX_TLS_OPENING_RETRY;
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_OPENING;
+				}
+			}
+		}
+    }
+
+    if (result != 0)
+    {
+        if (on_io_open_complete != NULL)
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_002: [ The tlsio_appleios shall report the open operation status using the IO_OPEN_RESULT enumerator defined in the `xio.h`.]*/
+            /* Codes_SRS_TLSIO_APPLEIOS_32_039: [ If the tlsio_appleios_open failed to open the tls connection, and the on_io_open_complete callback was provided, it shall call the on_io_open_complete with IO_OPEN_ERROR. ]*/
+            (void)on_io_open_complete(on_io_open_complete_context, IO_OPEN_ERROR);
+        }
```

Comment:
open_complete shall not be called if open immediately failed. The convention is that the callback is not expected and it can do more damage than good.